raiden/network/transport/matrix/transport.py

```diff
@@ -256,16 +265,27 @@ def _run(self) -> None:  # type: ignore
             f"node:{to_checksum_address(self.transport._raiden_service.address)} "
             f"recipient:{to_checksum_address(self.receiver)}"
         )
-        # run while transport parent is running
+        # run while transport parent is running and we're not idle
         while not self.transport._stop_event.ready():
             # once entered the critical section, block any other enqueue or notify attempt
             with self._lock:
                 self._notify_event.clear()
                 if self._message_queue:
+                    self._idle_since = 0
                     self._check_and_send()
+                else:
+                    self._idle_since += 1
+
+            if self.is_idle:
+                self.log.debug("Exiting idle RetryQueue", queue=self)
+                return
```

Comment:
this doesn't look right to me, as far as I know there is no place to *restart* the RetryQueue, if our partner node goes offline, the number of retries is reached, the queue is considered idle and then exits, once the partner comes back online we won't be able to communicate with it.

Reply:
They are never restarted. They only become idle if their message queue is empty for the specified amount of iterations. 
If then a message needs to be sent to the same peer `MatrixTransport._get_retrier()` will instantiate a new object.
I'll clarify the comments and also add a test for the negative case.

Reply:
Ahh, I see, I missed the check for `_message_queue`. Why not clearing as soon as it is empty?

On a side note: What about synchronization, this should only be done if there is no thread concurrently adding a message, is the `self._lock` being acquired in the `MatrixTransport`? (Humm, I don't really like this, having a lock being acquired outside of the owner class is usually a pain to understand)

Reply:
The lock is acquired in `RetryQueue._run` (and I added an assert in this PR inside `_check_and_send` to ensure that's the case).

`MatrixTransport._get_retrier()` is, in the case we're talking about, just a dict lookup.
And in the two places it's called, the `enqueue()` method is called immediately afterwards. Therefore there can't be a gevent context switch.

Of course that could change. But yeah holding the lock from the outside seems wrong...

Reply:
Not really, `_get_retrier` will also initialize a new greenlet, meaning it will be schedule and I guess it could context-switch.

Did you measure the impact of this change? Is it really necessary? I have the impression this will introduce hard to debug bugs.

Another approach is to simply wait on the `_notify_event` *without* a timeout if there are no messages in the queue, then there will be no race conditions.